Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+            || (MkextDecompress && NewOffset > OutBufferSize)) {
+            XmlDocumentFree (PlistXml);
+            FreePool (PlistBuffer);
+            if (MkextDecompress) {
+              FreePool (BinaryOffsetStrings);
+            }
+            return EFI_BUFFER_TOO_SMALL;
+          }
+
+          if (MkextDecompress) {
+            MkextOutExecutableEntry                 = (MKEXT_V2_FILE_ENTRY *) &OutBuffer[CurrentOffset];
+            MkextOutExecutableEntry->CompressedSize = 0;
+            MkextOutExecutableEntry->FullSize       = SwapBytes32 (BinFullSize);
+
+            //
+            // Compressed size == 0 means no compression.
+            //
+            if (BinCompSize > 0) {
+              if (OcOverflowAddU32 (BinOffset, BinCompSize, &Tmp)
+                || Tmp > MkextSize - sizeof (MKEXT_V2_FILE_ENTRY)
+                || DecompressZLIB (
+                  MkextOutExecutableEntry->Data,
+                  BinFullSize,
+                  MkextExecutableEntry->Data,
+                  BinCompSize
+                ) != BinFullSize) {
+                XmlDocumentFree (PlistXml);
+                FreePool (PlistBuffer);
+                FreePool (BinaryOffsetStrings);
+                return EFI_INVALID_PARAMETER;
+              }
+            } else {
+              if (OcOverflowAddU32 (BinOffset, BinFullSize, &Tmp)
+                || Tmp > MkextSize - sizeof (MKEXT_V2_FILE_ENTRY)) {
+                XmlDocumentFree (PlistXml);
+                FreePool (PlistBuffer);
+                FreePool (BinaryOffsetStrings);
+                return EFI_INVALID_PARAMETER;
+              }
+              CopyMem (MkextOutExecutableEntry->Data, MkextExecutableEntry->Data, BinFullSize);
+            }
+
+            if (!AsciiUint64ToLowerHex (
+              &BinaryOffsetStrings[Index * MKEXT_OFFSET_STR_LEN],
+              MKEXT_OFFSET_STR_LEN,
+              CurrentOffset
+              )) {
+              XmlDocumentFree (PlistXml);
+              FreePool (PlistBuffer);
+              FreePool (BinaryOffsetStrings);
+              return EFI_INVALID_PARAMETER;
+            }
+            XmlNodeChangeContent (BundleExecutable, &BinaryOffsetStrings[Index * MKEXT_OFFSET_STR_LEN]);
+          }
+
+          //
+          // Move to next bundle.
+          //
+          CurrentOffset = NewOffset;
+          break;
+        }
+      }
+    }
+
+    if (MkextDecompress) {
+      PlistFullSize = UpdateMkextV2Plist (&MkextHeaderOut->V2, PlistXml, CurrentOffset);
+    }
+    XmlDocumentFree (PlistXml);
+    FreePool (PlistBuffer);
+    if (MkextDecompress) {
+      FreePool (BinaryOffsetStrings);
+
+      if (PlistFullSize == 0
+        || OcOverflowAddU32 (CurrentOffset, PlistFullSize, OutMkextSize)) {
+        return EFI_INVALID_PARAMETER;
+      }
+    } else {
+      //
+      // Account for plist, 128 bytes per bundle for future plist expansion, 
```

Comment:
Why is 128 bytes enough? Provide some brief explanation, and make it a named constant.